### PR TITLE
Ada AC2 Ocelots Class - Elaine(SiOk) Sohn

### PR DESCRIPTION
**Pull request by sioksohn**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -117,14 +117,16 @@ def test_moves_movie_from_watchlist_to_empty_watched():
...
+    assert len(updated_data["watched"]) == 1    
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
-
-@pytest.mark.skip()
+    assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1
+    assert updated_data["watched"][0]["genre"] == GENRE_1
+    assert updated_data["watched"][0]["rating"] == RATING_1
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:127]: Nice assertions for all the relevant movie data!

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -141,14 +143,17 @@ def test_moves_movie_from_watchlist_to_watched():
...
-    
-    raise Exception("Test needs to be completed.")
+    assert len(updated_data["watched"]) == 2   
+        
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    assert movie_to_watch not in updated_data["watchlist"]
+    assert movie_to_watch in updated_data["watched"]
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:153]: Great use of `movie_to_watch` to confirm the object we set up in the `# Arrange` step is in the `watched` list and was removed from `watchlist`!

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,129 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if title == None or genre == None or rating == None:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:4]: Nice handling to return `None` for missing data! In Python, when we test if a variable is pointing to `None` the PEP8 style guide recommends using `is` or `is not` for comparison rather than operators like `==`. 

`None` is a special object in Python, it’s a `singleton`, meaning there is only ever one instance of `None` in memory. All variables holding `None` are actually pointing to the same instance. By using `is` to compare against `None`, we avoid issues that could be caused by a class’s custom implementation of the comparison operators, and we gain some speed over a check using `==` since we are checking against identity rather than looking up if the variable's held type has a comparison function that should be used.
```python
if title is None or genre is None or rating is None :
    return None
```
If you’d like more info on `is` vs `==` for `None`, here’s a couple resources to get started =]
https://peps.python.org/pep-0008/#programming-recommendations
http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html


Something else to consider: here we're explicitly checking if the values are None, but what if we got an empty string for `movie_title`? If we wanted to check for empty strings at the same time as we check for None, we could rely on the truthy/falsy values of the parameters and write something like: 
```python
if not title or not genre or not rating:
    return None
```

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,129 @@
...
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:20]: This function works, but it contains an "anti-pattern" that is important to be aware of.  An element of the list that is being iterated over is being removed inside the for loop. This won't cause problems in this particular case, but consider this slightly different implementation:
```
def watch_movie(user_data, title):
    for i in range(len(user_data["watchlist"])):
        item = user_data["watchlist"][i]

        if item["title"] == title:
            user_data["watched"].append(item)
            user_data["watchlist"].remove(item)

    return user_data
```
If we were to test this in a scenario where `user_data["watchlist"]` has 2 elements, and the first element is the movie that we are moving to `user_data["watched"]`, during the second iteration through the loop it will throw an IndexError, because we will try to access the 2nd element in `user_data["watchlist"]`, but there no longer is a second element.  

This is a coding pattern that I recommend avoiding because it has so much potential for unforeseen errors. One fix is to return immediately after making the swap (ie, add `return user_data` inside the if-block so that the loop ends as soon as the swap is made).  Another fix is to use the loop to find the item, but not make the swap until after the loop ends.

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,129 @@
...
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    count, total_rating = 0, 0.0
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:27]: I would consider removing `count`. We can build up that value over the loop, but we can get that same value by calling `len(user_data["watched"])`.

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,129 @@
...
+    for movie in user_data["watched"]:
+        count += 1
+        total_rating += movie["rating"]
+    return total_rating / count if count != 0 else 0.0
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:31]: Neat use of a python ternary operator! 

This is a little bit personal taste, but to make the code even clearer to readers, I recommend changing this around to make the validation check a guard clause at the top of the function. Right now we allocate space for `count` and `total_rating` before we validated if there are ratings to average. If we do the validation check first, then we only allocate space for variables that will be used.
```python
if not user_data["watched"]:
    return 0.0

count, total_rating = 0, 0.0
...
```

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,129 @@
...
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:26]: Nice approach! Another way we could do this is using a list comprehension and the `sum` function:

```python
# The list comprehension inside the `sum` call pulls the value of the 
# `rating` key out of each item in the list `user_data['watched']`
total = sum(movie["rating"] for movie in user_data["watched"])
avg_rating = total / len(user_data["watched"])
return avg_rating
```

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,129 @@
...
+    count, total_rating = 0, 0.0
+    for movie in user_data["watched"]:
+        count += 1
+        total_rating += movie["rating"]
+    return total_rating / count if count != 0 else 0.0
 
+def get_most_watched_genre(user_data):
+    # if user_data["watched"] == []: return None
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:34]: We want to remove commented out code before opening PRs, especially in shared projects, so that we don't bring things into the codebase that could be confusing down the road. If we have been making commits as we work, we can always check out the code at those commits if we want to take a look at something we've since changed or deleted.

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,129 @@
...
+    genres = {}
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre not in genres:
+            genres[genre] = 1
+        else:
+            genres[genre] += 1
+    sorted_genres = sorted(genres.items(), key=lambda x:x[1], reverse=True)
+    return sorted_genres[0][0] if len(sorted_genres) > 0 else None
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:43]: Great use of a frequency map `genres` to count the genre occurrences! Another way we could get the maximum value out of `genres` is with the `max` function:
```python
max_value = max(genres, key=genres.get)
return max_value
```

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,129 @@
...
+        for friend_movie in friend_dict["watched"]:
+            friend_movies.add(friend_movie["title"])
+    unique_title = user_movies - friend_movies
+
+    unique_movies = []
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_title and movie not in unique_movies:
+            unique_movies.append(movie)
+
+    return unique_movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:64]: Really neat approach using sets! 

Something to think about, here we will create 3 sets and a list, and we will iterate over `user_data["watched"]` twice and all data in `user_data["friends"]` once.

If we only create a set of the friend's titles, we can get the benefits of quick look up using a set, and make a single pass over `user_data["watched"]`.
```python
friend_movies = set()
for friend in user_data["friends"]:
    for movie in friend["watched"]:
        friend_movies.add(movie["title"])

unique_movies = []
for movie in user_data["watched"]:
    if movie["title"] not in friends_movie_set:
        unique_movies.append(movie)

return unique_movies
```

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,129 @@
...
+
+    # if user_data["watched"] == []: return []
+    
+    streamings = set()
+    for service in user_data["subscriptions"]:
+        streamings.add(service)
+
+    recommended = []
+    for friend in user_data["friends"]:
+            for movie in friend["watched"]:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:100]: It looks like the indentation is too deep in this for loop, you might have indented twice.

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,129 @@
...
+                if movie in get_friends_unique_watched(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:101]: Love to see the code reuse!

This implementation will call `get_friends_unique_watched` in every iteration of the loop. Since our input data, and thus the result of calling `get_friends_unique_watched`, shouldn't be changing, we could reduce the processing time of our function by creating a variable that holds the result of  `get_friends_unique_watched` before the loop starts. Because `get_friends_unique_watched` loops over our friend data already and provides us a list of movies that our user hasn't seen, we could also remove the for loops through the friend data here:

```python
unique_friend_movies = get_friends_unique_watched(user_data)
recommended = []

for movie in unique_friend_movies:
    if movie['host'] in user_data['subscriptions']:
        recommended.append(movie)

return recommended
```

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,129 @@
...
+            rec_by_genre.append(movie)
+    return rec_by_genre
+
+def get_rec_from_favorites(user_data):
+    rec_movies = []
+    for fav_movie in user_data["favorites"]:
+        for uniq_movie in get_unique_watched(user_data):
+            if fav_movie["title"] == uniq_movie["title"]:
+                rec_movies.append(fav_movie)
+    return rec_movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:125]: We're nesting loops a little more than required here. We could remove the loop over `user_data["favorites"]` and keep the loop over the result of `get_unique_watched`. Then in each iteration, we'd check if the current move is in `user_data['favorites']`.
```python
for uniq_movie in get_unique_watched(user_data):
    if uniq_movie in user_data['favorites']:
        rec_movies.append(uniq_movie)

return rec_movies
```

**Review by kelsey-steven-ada — COMMENTED** [review]
🎉 Congrats on completing your first project at Ada! 🎉 I’ve added some suggestions & questions, please let me know if there's anything I can clarify.
